### ISSUE TITLE
custom-elements-es5-adapter should not prevent usage of ES6 classes

### ISSUE THREAD

**Issue by MajorBreakfast** [webcomponents/webcomponentsjs#871]
Currently the `custom-elements-es5-adapter` prevents you from using customElement.define() with a real ES6 class. What you get then is a really unhelpful error message (class must be called with `new`, bla bla).

https://github.com/webcomponents/custom-elements/blob/master/src/native-shim.js#L120

I propose:
- ES5 class -> Apply shim
- ES6 class -> No shim (use native impl directly), display a warning `'customElements.define() was called with an ES6 class: custom-elements-es5-adapter.js is intended for projects transpiled to ES5'`

Advantages:
- Still works with ES6 classes for debugging
- Reminds developers to transpile all their files

I had a project where the `node_modules` folder was excluded from transpilation. It took me quite some time to figure it out, because there was no helpful error message.

**Comment by MajorBreakfast** [comment]
```
function isClass(func) {
  return typeof func === 'function' 
    && Function.prototype.toString.call(func).slice(0, 5) === 'class';
}
```

**Comment by CITguy** [comment]
This might have been fixed with the dependency update to `@webcomponents/custom-elements` version `1.2.0` (988d4a149f5049b9fcf46b5a34528c9538d1acd7). 

The adapter was modified to replace the ironically used ES6 class syntax with a `Reflect`-based solution ([`fea35b08`](https://github.com/webcomponents/custom-elements/commit/fea35b08cb989f3f0a21b0ade8aa0f15f93bc49d)).

The following should work as of `@webcomponents/webcomponents` versions `v1.2.4` and `v2.0.4`.

## TL;DR
```html
<head>
  <!-- ... -->
  <link rel="stylesheet" href="path/to/application-styles.css" />

  <script src="path/to/node_modules/@webcomponents/webcomponentsjs/custom-elements-es5-adapter.js"></script>
  <script src="path/to/node_modules/@webcomponents/webcomponentsjs/webcomponents-loader.js"></script>
  <!-- polyfills for the current browser get injected here (if needed) -->
</head>
<body>
  <!-- ... -->

  <script src="path/to/application-logic.js"></script>
</body>
```

## Updates to your markup

Instead of the `<span>` hack:

```html
<span id="ce-es5-adapter">	
  <script>	
    if (!window.customElements) {	
      var elAdapter = document.querySelector('#ce-es5-adapter');	
      elAdapter.parentElement.removeChild(elAdapter);	
    }	
  </script>	
  <script src="path/to/custom-elements-es5-adapter.js"></script>	
</span>
```

You can just load the adapter directly:

```html
<script src="path/to/custom-elements-es5-adapter.js"></script>
```

**Comment by TimvdLippe** [comment]
As @CITguy pointed out, this should now be fixed. Please reopen if you are still encountering issues.

**Comment by sebagr** [comment]
any chance this is happening again? I'm on 1.2.6 and I'm getting "Class constructor Foo cannot be invoked without 'new'".

**Comment by CITguy** [comment]
@sebagr: are you loading the adapter before loading your own code?

**Comment by sebagr** [comment]
I'm loading webcomponent-loader.js before my app.html import. Since the
loader loads whatever's needed on demand, can it be the case that my app
gets loaded first?

For the time being I switched to loading webcomponents-lite to see if that
helps get rid of the problem.

**Comment by CITguy** [comment]
The loader and the adapter are two different files. If your code is compiled to ES5, you have to load the _adapter_ before your code loads.

**Comment by sebagr** [comment]
mmm I thought it was the loader which was loading the adapter in turn...
are you saying it is not?

I'm working on a Polymer project, maybe it's Polymer that's including the
adapter then? I'm not sure how to see what's loading it :(

**Comment by sebagr** [comment]
Thank you for your detailed answer! I'm currently serving ES5 only for Googlebot, as it doesn't understand ES6 just yet. All other user agents are getting the ES6 version. Both bundles are created by  Polymer and up until now I thought it was webcomponents-loader that was injecting the adapter based on feature detection. Now I understand this is _not_ the case. Am I correct? If so, then there's nothing to worry about here.
